flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/DispatcherJob.java

```diff
@@ -0,0 +1,242 @@
+					}
+				} else { // failure during initialization
+					if (submissionType == SubmissionType.RECOVERY) {
+						jobResultFuture.completeExceptionally(throwable);
+					} else {
+						jobResultFuture.complete(ArchivedExecutionGraph.createFromInitializingJob(
+							jobId,
+							jobName,
+							throwable,
+							JobStatus.FAILED,
+							initializationTimestamp));
+					}
+				}
+			}
+			return null;
+		}));
+	}
+
+	public CompletableFuture<ArchivedExecutionGraph> getResultFuture() {
+		return jobResultFuture;
+	}
+
+	public CompletableFuture<JobDetails> requestJobDetails(Time timeout) {
+		synchronized (lock) {
+			if (isRunning()) {
+				return getJobMasterGateway().thenCompose(jobMasterGateway -> jobMasterGateway.requestJobDetails(
+					timeout));
+			} else {
+				int[] tasksPerState = new int[ExecutionState.values().length];
+				return CompletableFuture.completedFuture(new JobDetails(
+					jobId,
+					jobName,
+					initializationTimestamp,
+					0,
+					0,
+					jobStatus,
+					0,
+					tasksPerState,
+					0));
+			}
+		}
+	}
+
+	public CompletableFuture<Acknowledge> cancel(Time timeout) {
+		synchronized (lock) {
+			if (isRunning()) {
+				return getJobMasterGateway().thenCompose(jobMasterGateway -> jobMasterGateway.cancel(timeout));
+			} else {
+				log.info("Cancellation during initialization requested for job {}. Job will be cancelled once JobManager has been initialized.", jobId);
+				jobStatus = JobStatus.CANCELLING;
+				return jobResultFuture.thenApply(ignored -> Acknowledge.get());
+			}
+		}
+	}
+
+	public CompletableFuture<JobStatus> requestJobStatus(Time timeout) {
+		synchronized (lock) {
+			if (isRunning()) {
+				return getJobMasterGateway().thenCompose(jobMasterGateway -> jobMasterGateway.requestJobStatus(timeout));
+			} else {
+				return CompletableFuture.completedFuture(jobStatus);
+			}
+		}
+	}
+
+	public CompletableFuture<ArchivedExecutionGraph> requestJob(Time timeout) {
+		synchronized (lock) {
+			if (isRunning()) {
+				return getJobMasterGateway().thenCompose(jobMasterGateway -> jobMasterGateway.requestJob(timeout));
+			} else {
+				return CompletableFuture.supplyAsync(() -> ArchivedExecutionGraph.createFromInitializingJob(jobId, jobName, null, jobStatus, initializationTimestamp));
+			}
+		}
+	}
+
+	public boolean isRunning() {
+		synchronized (lock) {
+			return jobStatus == JobStatus.RUNNING;
+		}
+	}
+
+	public CompletableFuture<JobMasterGateway> getJobMasterGateway() {
+		Preconditions.checkState(
+			isRunning(),
+			"JobMaster Gateway is not available during initialization");
+		try {
+			return jobManagerRunnerFuture.thenCompose(JobManagerRunner::getJobMasterGateway);
+		} catch (Throwable e) {
+			throw new IllegalStateException("JobMaster gateway is not available", e);
+		}
+	}
+
+	@Override
+	public CompletableFuture<Void> closeAsync() {
```

Comment:
Unless we add a new `JobStatus.SUSPENDING` or `JobStatus.TERMINATING`, I don't see a good way of reporting to the user that we are waiting for the initialization to be finished so that we can close the JobManagerRunner.

Afaik this PR is not changing the semantics of a job stuck INITIALIZING. Previously, a job would report `CREATED` or `RUNNING` until it reaches the final `SUSPENDED` state, if `Dispatcher.onStop` get's called.